### DISCUSSION TITLE
How to set up Leantime with caddy

### DISCUSSION THREAD

**Discussion by fyrak1s** [Leantime/leantime#624] [Q&A]
I am trying to set up Leantime behind [caddy](https://caddyserver.com/) (an nginx alternative). I am not using docker, but the normal packaged version.

However for some reason I am getting infinite redirects when I am entering `leantime.mydomain.com/install`.

I am getting an error like this: `ERR_TOO_MANY_REDIRECTS`

I had asked in the caddy community but they said the fault is in the configuration of the application, not caddy itself.

[Here](https://caddy.community/t/too-many-redirects-occurred-trying-to-open-site/6656) is somebody describing a similar issue. Not sure if it is the same problem, but I'm still leaving it here. (See first answer)

Any hint what might be wrong? I could not find anything in the leantime docs.

PS: Here is my caddy config for reference (not finished porting, but it should not have the infinite redirects):
```
leantime.mydomain.com {
	root * /var/www/html/leantime/public
	php_fastcgi unix//run/php-fpm/www.sock
	file_server
}
```

EDIT: Could it be due to the fact that caddy automatically employs ssl?

**Comment by marcelfolaron** [comment]
Did you try to set the app url in the config file? That is probably where I would start.

**Comment by fyrak1s** [comment]
When I leave `appUrl` emtpy my browser shows `ERR_TOO_MANY_REDIRECTS` when entering the url `leantime.domain.com/install/` in the browser.

When I set into `$appUrl = "leantime.domain.com"` my browser shows this infinitely in the url bar:
```
https://leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/.......................
```
And also `ERR_TOO_MANY_REDIRECTS` as the error message.

**Comment by marcelfolaron** [comment]
Can you try to set it including the protocol? http(s)://leantime.domain.com
My guess is that the infinite redirect is happening because of a http to
https redirect.

**Reply by fyrak1s** [reply]
Thank you. In that case I am getting:
```
This site can’t provide a secure connection
leantime.mydomain.com sent an invalid response.
ERR_SSL_PROTOCOL_ERROR
```

**Reply by fyrak1s** [reply]
Any help? I would really like to try out leantime, but I just can't if it doesnt work with caddy. I have asked in the caddy community and browsed the internet, but nobody manages to get it working.

**Reply by marcelfolaron** [reply]
That error message seems to indicate an error on the caddy site. Let me check a few things and get back to you within the hour

**Reply by marcelfolaron** [reply]
https://caddy.community/t/this-site-can-t-provide-a-secure-connection/5681/2

From what I can gather something is up with ssl. 
I assume you set the leantime config to be `https://leantime.domain.com` ?

My first thought is that there is something wrong with the certificate or the ports. Can you paste the current caddy config you are using?

**Reply by marcelfolaron** [reply]
Does your certificate include leantime.domain.com?

**Reply by fyrak1s** [reply]
Hi @marcelfolaron thank you for comming back to me.
I don't know what happened there. The error remains `ERR_TOO_MANY_REDIRECTS` (I dont know why the ssl error happened but its gone).

**To summarize:**
- with `appUrl=https://leantime.domain.com` the browser goes to `https://leantimedomain.com/install` but shows `ERR_TOO_MANY_REDIRECTS`.

- with `appUrl=leantime.domain.com` the browsers url bar shows `https://leantime.domain.com/install/leantime.domain.com/install/leantime.domain.com/install/.....` infinitely also with the error `ERR_TOO_MANY_REDIRECTS`

According to the browser inspector those are 302 redirects.

Btw caddy handles ssl and certificates automatically.

Are you sure it isn't related to something like this https://caddy.community/t/too-many-redirects-occurred-trying-to-open-site/6656/2 ?

There is also the planka project (a kanboard) which has a specific `TrustProxy` setting, for reverse proxies that handle ssl automatically https://github.com/plankanban/planka/issues/163. Not sure if that helps here.

**Comment by marcelfolaron** [comment]
I see. I didn't realize caddy didn't read graces files. Basically you would
have to take the htaccess redirect rules and create caddy equivalents. I am
looking through the syntax to see how hard that would be

**Reply by fyrak1s** [reply]
Hey! Yes in the caddy discord, they said that apparently "modern" php applications have those definitions inside index.php, which is not the case for leantime sadly. And thus probably it needs to be translated. Thank you for looking into this!

**Reply by fyrak1s** [reply]
@marcelfolaron any update on this so far?

**Comment by keskinonur** [comment]
I can mitigate general problems with the following Caddyfile, but it is not perfect:

```
lean.HOSTNAME.TLD {
    tls internal
    root * /PATH/lean/public
    file_server
    log {
        output file /PATH/lean/log/access.log
    }

    rewrite /install /index.php?install=true
    rewrite / /index.php?act=dashboard.show
    rewrite /resetPassword /index.php?resetPassword=true
    rewrite /update /index.php?update=true

    @abc path_regexp twopath ^/([^/\.]+)/([^/\.]+)/?$
    rewrite @abc /index.php?act={re.twopath.1}.{re.twopath.2}

    @xyz path_regexp mypath ^/([^/\.]+)/([^/\.]+)/([^/\.]+)/?$
    rewrite @xyz /index.php?act={re.mypath.1}.{re.mypath.2}&id={re.mypath.3}

    encode zstd gzip
    header Strict-Transport-Security max-age=31536000;
    php_fastcgi unix//run/php/php7.4-fpm.sock
}
```
✅ Remarks:
- **tls internal** is for Cloudflare's SSL proxy
- direct rewrite lines are self explanatory
- named and path matchers for route redirects

⁉️ Remaining issues:

1. Onboarding modal box comes blank screen, help/showOnboardingDialog?module=newProject because of unhandled query string at the end.
2. /users/editUser/XX/ not working

If some could continue where I left off, that would be great.
Thanks 👋

**Reply by fyrak1s** [reply]
Thanks a lot @sudOnur !!

@marcelfolaron Would you be able to help us out on the last mile?

**Comment by marcelfolaron** [comment]
This is awesome!

#2 is solved in master. Just needs to be released. I promise I am working
on it.

#1 which parameter is not handled? Do you get a 404 or is there an error?

**Reply by fyrak1s** [reply]
So would the fix be included in Leantime v2.1.9 ?

I will try that version when I get the time and report back here! :)
> #1 which parameter is not handled? Do you get a 404 or is there an error?

@sudOnur ^

**Comment by kevinpetit** [comment]
A late reply, but I've been testing and working on getting Leantime to work properly behind Caddy, and this is the configuration file I was able to build which works:

```
        rewrite /install /index.php?install=true
        rewrite / /index.php?act=dashboard.show
        rewrite /resetPassword /index.php?resetPassword=true
        rewrite /update /index.php?update=true

        @abc path_regexp twopath ^/([^/\.]+)/([^/\.?]+)/?$
        rewrite @abc /index.php?act={re.twopath.1}.{re.twopath.2}&{query}

        @xyz path_regexp mypath ^/([^/\.]+)/([^/\.]+)/([^/\.]+)/?$
        rewrite @xyz /index.php?act={re.mypath.1}.{re.mypath.2}&id={re.mypath.3}

```
The important part here is that the query string needs to be passed through, once I did that all uploads show up. This also fixes the modals.

**Reply by autisimo** [reply]
you are a legend, i have been frustrated for weeks to get this working. one thing i had to do to get it working was to do the setup and everything using apache, and then switching to caddy with the config you provided

**Reply by kevinpetit** [reply]
You can do the setup without Apache (if it doesn't work with Caddy), but it requires the command line.
As usual, complete your .env file with the right values. Once that is done, go to the CLI and go to your install folder's bin folder.
Then perform `./leantime db:install` which will perform the database installation, and after that use `./leantime user:add` to create your user.
That should allow you to skip the installer and have a working installation without using Apache at all.
